File: codewars/6kyu/whos-likes-it/dduck-whos-likes-it.cpp
// https://www.codewars.com/kata/5266876b8f4bf2da9b000362/
//
// $ g++ -std=c++11 space.cpp

#include <iostream>

#include <string>
#include <vector>

using namespace std;

std::string likes(const std::vector <std::string> &names) {
    switch (names.size()) {
        case 0:
            return "no one likes this";
        case 1:
            return std::string(names[0]) + " likes this";
        case 2:
            return std::string(names[0]) + " and " + names[1] + " like this";
        case 3:
            return std::string(names[0]) + ", " + names[1] + " and " + names[2] + " like this";
        default:
            return std::string(names[0]) + ", " + names[1] + " and " + std::to_string(names.size() - 2) +
                   " others like this";
    }
}

int main() {

    std::cout << likes({}) << "no one likes this" << "\n";
    std::cout << likes({"Peter"}) << "Peter likes this" << "\n";
    std::cout << likes({"Jacob", "Alex"}) << "Jacob and Alex like this" << "\n";
    std::cout << likes({"Max", "John", "Mark"}) << "Max, John and Mark like this" << "\n";
    std::cout << likes({"Alex", "Jacob", "Mark", "Max"}) << "Alex, Jacob and 2 others like this" << "\n";
    return 0;
}

File: codewars/6kyu/whos-likes-it/dduck-whos-likes-it.py
# https://www.codewars.com/kata/5266876b8f4bf2da9b000362/

from unittest import TestCase


def likes(input):
    switcher = {
        0: lambda names: "no one likes this",
        1: lambda names: f"{names[0]} likes this",
        2: lambda names: f"{names[0]} and {names[1]} like this",
        3: lambda names: f"{names[0]}, {names[1]} and {names[2]} like this",
    }
    default = lambda names: f"{names[0]}, {names[1]} and {len(names) - 2} others like this"

    return switcher.get(len(input), default)(input)


TestCase().assertEqual(likes([]), 'no one likes this')
TestCase().assertEqual(likes(['Peter']), 'Peter likes this')
TestCase().assertEqual(likes(['Jacob', 'Alex']), 'Jacob and Alex like this')
TestCase().assertEqual(likes(['Max', 'John', 'Mark']), 'Max, John and Mark like this')
TestCase().assertEqual(likes(['Alex', 'Jacob', 'Mark', 'Max']), 'Alex, Jacob and 2 others like this')
